File: 19_remove_nth_node_from_end_of_list/19_remove_nth_node_from_end_of_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        length = 0
        current = head
        while current:
            length += 1
            current = current.next
        
        # 步驟2：計算要移除的節點的位置
        target_index = length - n
        
        # 如果要移除的是頭節點
        if target_index == 0:
            return head.next
        
        # 步驟3：第二次遍歷鏈表，找到目標節點的前一個節點
        current = head
        for i in range(target_index - 1):
            current = current.next
        
        # 步驟4：移除節點
        current.next = current.next.next
        
        # 步驟5：返回修改後的鏈表的頭節點
        return head

File: 19_remove_nth_node_from_end_of_list/19_remove_nth_node_from_end_of_list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function (head, n) {
  var length = 0;
  var current = head;
  while (current) {
    current = current.next;
    length += 1;
  }
  target = length - n;
  if (target == 0) return head.next;
  current = head;
  for (i = 0; i < target - 1; i++) {
    current = current.next;
  }
  current.next = current.next.next;
  return head;
};
